### DISCUSSION TITLE
Custom Transport: Why is a clone required for the RunspaceConnectionInfo?

### DISCUSSION THREAD

**Discussion by JustinGrote** [PowerShell/PowerShell#19661] [Q&A] [answered]
I was just curious why this is required:
https://github.com/PowerShell/PowerShell/blob/c444645b0941d73dc769f0bba6ab70d317bd51a9/src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs#L355-L362

It looks like connection info should be fairly static and immutable, so why clone it?
@PaulHigin?

**Comment by rhubarb-geek-nz** [comment]
> It looks like connection info should be fairly static and immutable, so why clone it?

The object has set methods on it so it isn't either static or immutable. Just an idea, you could take an existing one, clone it and then change the timeouts, or culture, or any of the settable properties.

**Reply by JustinGrote** [reply]
I just meant in terms of why can't the existing object be used, why does it need to be cloned.

**Comment by rhubarb-geek-nz** [comment, marked as answer]
Just because an object has a Clone method does not mean you have to clone it to use it.

The [IClonable](https://learn.microsoft.com/en-us/dotnet/api/system.icloneable?view=net-7.0) is very relaxed on what it all means in any situation.

```
It does not specify whether the cloning operation performs a deep copy, a shallow copy, or something in between. Nor does it require all property values of the original instance to be copied to the new instance.
```

**Reply by JustinGrote** [reply]
That's a fair point, let me try returning the original object and see if I hit any edge cases.

**Reply by JustinGrote** [reply]
But in @PaulHigin's example the clone was explicit too
https://github.com/PaulHigin/NamedPipeConnection#clone

**Comment by rhubarb-geek-nz** [comment]
It may be that PowerShell is internally calling Clone() so that it is doing its best to isolate connections from each other, so if another connection did start changing timeouts or whatever, the cloned versions would not be affected.
